src/ctia/schemas/utils.clj

```diff
@@ -18,55 +19,169 @@
+                                                      (throw (ex-info (str "Missing " service-kw " service function: "
+                                                                           (s/explicit-schema-key fn-kw))
+                                                                      {}))))))
+                                           fn-kws)]
+                     {(s/explicit-schema-key service-kw) service-fns})))))
+        selectors))
+
+(s/defn service-subgraph-from-schema :- (s/pred map?)
+  "Given a schema describing a Trapperkeeper graph,
+  returns just the elements in graph mentioned
+  in the schema as 'explicit keys' using service-subgraph.
+  Throws an exception if any levels are missing.
+  
+  (service-subgraph-from-schema
+    {:ConfigService {:get-in-config (fn [...] ...)
+                     :get-config (fn [...] ...)}
+     :FooService {:f1 (fn [...] ...)
+                  :f2 (fn [...] ...)}
+     :BarService {:b1 (fn [...] ...)}}
+    {:ConfigService {:get-in-config (s/=> ...)}
+     (s/optional-key :FooService) {:f1 (s/=> ...)
+                                   (s/optional-key :f2) (s/=> ...)}})
+  ;=> {:ConfigService {:get-in-config (fn [...] ...)}
+  ;    :FooService {:f1 (fn [...] ...)
+  ;                 :f2 (fn [...] ...)}}
+  "
+  [graph-value :- (s/pred map?)
+   schema :- (s/protocol s/Schema)]
+  {:pre [(map? schema)]}
+  (service-subgraph
+    graph-value
+    ;; this could be factored out a la schema.coerce/coercer for
+    ;; better performance.
+    (into {}
+          (map (fn [[service-kw service-fns]]
+                 {:pre [(map? service-fns)]}
+                 (when (s/specific-key? service-kw)
+                   {service-kw
+                    (into #{}
+                          (filter s/specific-key?)
+                          (keys service-fns))})))
+          schema)))
+
+(s/defn service-subschema :- (s/protocol s/Schema)
+  "Given a schema shaped like a Trapperkeeper service graph, selects
+  the specified services and their optionality from graph. Throws an
+  exception on selected service functions that don't occur in graph.
+  
+  (service-subschema
+    {:ConfigService {:get-in-config (s/=> ...)
+                     (s/optional-key :get-config) (s/=> ...)}
+     (s/optional-key :FooService) {:f1 (s/=> ...)
+                                   :f2 (s/=> ...)}
+     :BarService {:b1 (s/=> ...)}}
+    {:ConfigService #{:get-config}
+     :FooService #{:f2 :f3}
+     :MissingService #{:m1}})
+  ;=> {:ConfigService {(s/optional-keys :get-config) (s/=> ...)}
+  ;    (s/optional-key :FooService) {:f2 (s/=> ...)}}
+  "
+  [graph-schema :- (s/protocol s/Schema)
+   selectors :- {(s/pred simple-keyword?)
+                 #{(s/pred simple-keyword?)}}]
+  {:pre [(map? graph-schema)]}
+  (into {}
+        (map (fn [[service-kw fn-kws]]
+               (assert (keyword? service-kw)
+                       (pr-str service-kw))
+               (assert (set? fn-kws))
```

Comment:
aren't these assert a duplicate of the schemas?

Reply:
Yes, I made the schema more specific via `SimpleKeywordSpecificKey` and removed the asserts.